File: packages/physics-core/src/models/pointMass.ts
import { EPS } from '../constants';
import type { Vec2 } from '../types';
import { norm2, sub } from '../vector';

export function pointMassDeflection(theta: Vec2, center: Vec2, thetaE: number): Vec2 {
  const d = sub(theta, center);
  const r2 = Math.max(norm2(d), EPS * EPS);
  const f = (thetaE * thetaE) / r2;
  return [f * d[0], f * d[1]];
}

export function pointMassPotential(theta: Vec2, center: Vec2, thetaE: number): number {
  const d = sub(theta, center);
  return thetaE * thetaE * Math.log(Math.sqrt(Math.max(norm2(d), EPS * EPS)));
}

export function pointMassImagePositions1D(beta: number, thetaE: number): [number, number] {
  const disc = Math.sqrt(beta * beta + 4 * thetaE * thetaE);
  return [0.5 * (beta + disc), 0.5 * (beta - disc)];
}

export function pointMassMagnifications1D(beta: number, thetaE: number): [number, number] {
  const u = Math.max(Math.abs(beta / thetaE), EPS);
  const term = (u * u + 2) / (2 * u * Math.sqrt(u * u + 4));
  return [0.5 + term, 0.5 - term];
}

File: python/cosmiclens_validate/point_mass.py
from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass(frozen=True)
class PointMassResult:
    theta_plus: float
    theta_minus: float
    mu_plus: float
    mu_minus: float


def point_mass_images(beta: float, theta_e: float) -> tuple[float, float]:
    disc = math.sqrt(beta * beta + 4.0 * theta_e * theta_e)
    return 0.5 * (beta + disc), 0.5 * (beta - disc)


def point_mass_magnifications(beta: float, theta_e: float) -> tuple[float, float]:
    eps = 1e-15
    u = max(abs(beta / theta_e), eps)
    term = (u * u + 2.0) / (2.0 * u * math.sqrt(u * u + 4.0))
    return 0.5 + term, 0.5 - term


def validate(beta: float = 0.3, theta_e: float = 1.0) -> PointMassResult:
    theta_plus, theta_minus = point_mass_images(beta, theta_e)
    mu_plus, mu_minus = point_mass_magnifications(beta, theta_e)
    return PointMassResult(theta_plus, theta_minus, mu_plus, mu_minus)


if __name__ == "__main__":
    print(validate())
